File: src/Billing/Mediation/Summary/Cache/AbstractCache.cs
// Ported from legacy Mediation/Cache/AbstractCache.cs.
// KEPT VERBATIM: the in-memory Cache + the Inserted/Updated/Deleted change-tracking, and
// Insert/InsertWithKey/UpdateThroughCache/Delete with their exact guard checks.
// ADAPTED: WriteAllChanges/WriteInserts/WriteUpdates/WriteDeletes run the generated SQL through an
// ISqlExecutor on the single shared MySqlConnection — the legacy DbWriterWithAccurateCount stored proc is
// gone (the executor returns the affected count directly), but the legacy CollectionSegmenter batching IS
// reused via Billing.Mediation.Sql.BatchSqlWriter, so any number of rows write in fixed-size segments. The
// legacy StaticExtInsertColumnParsedDic header is passed in as InsertHeader.
#nullable disable
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Billing.Mediation.Sql;
using MediationModel;

namespace TelcobrightMediation
{
    public abstract class AbstractCache<TEntity, TKey> where TEntity : ICacheble<TEntity>
    {
        private readonly object locker = new object();

        protected AbstractCache(
            Func<TEntity, TKey> dictionaryKeyGenerator,
            Func<TEntity, StringBuilder> insertCommandGenerator,
            Func<TEntity, StringBuilder> updateCommandGenerator,
            Func<TEntity, StringBuilder> deleteCommandGenerator,
            string insertHeader)
        {
            this.DictionaryKeyGenerator = dictionaryKeyGenerator;
            this.InsertCommandGenerator = insertCommandGenerator;
            this.UpdateCommandGenerator = updateCommandGenerator;
            this.DeleteCommandGenerator = deleteCommandGenerator;
            this.InsertHeader = insertHeader;
        }

        public Func<TEntity, TKey> DictionaryKeyGenerator { get; protected set; }
        public Func<TEntity, StringBuilder> InsertCommandGenerator { get; protected set; }
        public Func<TEntity, StringBuilder> UpdateCommandGenerator { get; protected set; }
        public Func<TEntity, StringBuilder> DeleteCommandGenerator { get; protected set; }

        /// <summary>"insert into &lt;table&gt; (col1,...) values " — replaces the legacy StaticExtInsertColumnParsedDic.</summary>
        protected string InsertHeader { get; }
        public virtual string EntityOrTableName => typeof(TEntity).Name;

        protected ConcurrentDictionary<TKey, TEntity> Cache { get; } = new ConcurrentDictionary<TKey, TEntity>();
        protected readonly ConcurrentDictionary<TKey, TEntity> InsertedItems = new ConcurrentDictionary<TKey, TEntity>();
        protected readonly ConcurrentDictionary<TKey, TEntity> UpdatedItems = new ConcurrentDictionary<TKey, TEntity>();
        protected readonly ConcurrentDictionary<TKey, TEntity> DeletedItems = new ConcurrentDictionary<TKey, TEntity>();

        public bool Exists(TKey key) => this.Cache.ContainsKey(key);

        public virtual CachedItem<TKey, TEntity> GetItemByKey(TKey key)
        {
            this.Cache.TryGetValue(key, out var selectedItem);
            return selectedItem == null ? null : new CachedItem<TKey, TEntity>(key, selectedItem);
        }

        public virtual List<TEntity> GetItems() => this.Cache.Values.ToList();
        public virtual List<TEntity> GetInsertedItems() => this.InsertedItems.Values.ToList();
        public virtual List<TEntity> GetUpdatedItems() => this.UpdatedItems.Values.ToList();
        public virtual List<TEntity> GetDeletedItems() => this.DeletedItems.Values.ToList();
        public int NumberOfInsertedItems => this.InsertedItems.Count;
        public int NumberOfUpdatedItems => this.UpdatedItems.Count;
        public int NumberOfDeletedItems => this.DeletedItems.Count;

        /// <summary>Seed the cache with an EXISTING (already-persisted) row — the PopulatePrevSummary entry.
        /// It goes into the cache only, NOT InsertedItems, so a later merge marks it Updated (not Inserted).</summary>
        public virtual void PopulateExisting(TEntity existing)
        {
            lock (this.locker)
            {
                var key = this.DictionaryKeyGenerator(existing);
                if (this.Cache.TryAdd(key, existing) == false)
                    throw new Exception("Could not add existing item to cache while populating, probably duplicate item.");
            }
        }

        public virtual void Delete(TKey key)
        {
            lock (this.locker)
            {
                this.Cache.TryGetValue(key, out var toBeDeleted);
                if (toBeDeleted == null) throw new Exception("Item does not exist in cache.");
                if (this.DeletedItems.ContainsKey(key) == false)
                {
                    if (this.DeletedItems.TryAdd(key, toBeDeleted) == false)
                        throw new Exception("Could not add item to deleted items.");
                }
                this.Cache.TryRemove(this.DictionaryKeyGenerator(toBeDeleted), out _);
            }
        }

        public virtual void UpdateThroughCache(TKey key, Action<TEntity> entityUpdater)
        {
            lock (this.locker)
            {
                if (this.DeletedItems.ContainsKey(key) == true)
                    throw new Exception("Item to be updated already exists in deleted items which is not allowed");
                this.Cache.TryGetValue(key, out var entityToBeUpdated);
                if (entityToBeUpdated == null) throw new Exception("Item does not exist in cache.");
                entityUpdater.Invoke(entityToBeUpdated);
                if (this.InsertedItems.ContainsKey(key) == false && this.UpdatedItems.ContainsKey(key) == false)
                {
                    if (this.UpdatedItems.TryAdd(key, entityToBeUpdated) == false)
                        throw new Exception("Could not add item to updated items, probably duplicate item.");
                }
            }
        }

        public virtual CachedItem<TKey, TEntity> Insert(TEntity newItem, Func<TEntity, bool> pkValidationMethod)
        {
            lock (this.locker)
            {
                TKey key = this.DictionaryKeyGenerator(newItem);
                return InsertWithKey(newItem, key, pkValidationMethod);
            }
        }

        public virtual CachedItem<TKey, TEntity> InsertWithKey(TEntity newItem, TKey key, Func<TEntity, bool> pkValidationMethod)
        {
            lock (this.locker)
            {
                if (pkValidationMethod.Invoke(newItem) == false)
                    throw new Exception("Primary key is either missing or invalid.");
                if (this.UpdatedItems.ContainsKey(key) == true)
                    throw new Exception("Item to be inserted already exists in updated items which is not allowed.");
                if (this.DeletedItems.ContainsKey(key) == true)
                    throw new Exception("Item to be inserted already exists in deleted items which is not allowed.");
                if (this.Cache.TryAdd(key, newItem) == false)
                    throw new Exception("Could not add item to cache, probably duplicate item.");
                if (this.InsertedItems.TryAdd(key, newItem) == false)
                    throw new Exception("Could not add item to inserted items, probably duplicate item.");
                return new CachedItem<TKey, TEntity>(key, newItem);
            }
        }

        // ---- write engine: thin executor over the single shared connection ----

        public virtual void WriteAllChanges(ISqlExecutor executor, int segmentSize = BatchSqlWriter.DefaultSegmentSize)
        {
            lock (this.locker)
            {
                WriteDeletes(executor, segmentSize);
                WriteInserts(executor, segmentSize);
                WriteUpdates(executor, segmentSize);
            }
        }

        public virtual int WriteInserts(ISqlExecutor executor, int segmentSize = BatchSqlWriter.DefaultSegmentSize)
        {
            lock (this.locker)
            {
                if (this.InsertedItems.Any() == false) return 0;
                if (this.InsertCommandGenerator == null) throw new Exception("InsertCommandGenerator is not set and null.");
                var values = this.GetInsertedItems().Select(c => this.InsertCommandGenerator(c)).ToList();
                int affected = BatchSqlWriter.WriteInsertsInSegments(executor, this.InsertHeader, values, segmentSize);
                this.InsertedItems.Clear();
                return affected;
            }
        }

        public virtual void WriteUpdates(ISqlExecutor executor, int segmentSize = BatchSqlWriter.DefaultSegmentSize)
        {
            lock (this.locker)
            {
                if (this.UpdatedItems.Any() == false) return;
                if (this.UpdateCommandGenerator == null) throw new Exception("UpdateCommandGenerator is null/not defined.");
                var statements = this.GetUpdatedItems().Select(c => this.UpdateCommandGenerator(c).ToString()).ToList();
                BatchSqlWriter.WriteStatementsInSegments(executor, statements, segmentSize);
                this.UpdatedItems.Clear();
            }
        }

        public virtual void WriteDeletes(ISqlExecutor executor, int segmentSize = BatchSqlWriter.DefaultSegmentSize)
        {
            lock (this.locker)
            {
                if (this.DeletedItems.Any() == false) return;
                if (this.DeleteCommandGenerator == null) throw new Exception("DeleteCommandGenerator is null/not defined.");
                var statements = this.GetDeletedItems().Select(c => this.DeleteCommandGenerator(c).ToString()).ToList();
                BatchSqlWriter.WriteStatementsInSegments(executor, statements, segmentSize);
                this.DeletedItems.Clear();
            }
        }
    }
}

File: java/src/main/java/com/telcobright/billing/mediation/summary/cache/AbstractCache.java
// Ported from legacy Mediation/Cache/AbstractCache.cs.
// KEPT VERBATIM: the in-memory Cache + the Inserted/Updated/Deleted change-tracking, and
// Insert/InsertWithKey/UpdateThroughCache/Delete with their exact guard checks.
// ADAPTED: WriteAllChanges/WriteInserts/WriteUpdates/WriteDeletes run the generated SQL through an
// ISqlExecutor on the single shared MySqlConnection — the legacy DbWriterWithAccurateCount stored proc is
// gone (the executor returns the affected count directly), but the legacy CollectionSegmenter batching IS
// reused via Billing.Mediation.Sql.BatchSqlWriter, so any number of rows write in fixed-size segments. The
// legacy StaticExtInsertColumnParsedDic header is passed in as InsertHeader.
package com.telcobright.billing.mediation.summary.cache;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import com.telcobright.billing.mediation.sql.BatchSqlWriter;
import com.telcobright.billing.mediation.sql.ISqlExecutor;

public abstract class AbstractCache<TEntity extends ICacheble<TEntity>, TKey> {
    private final Object locker = new Object();

    protected AbstractCache(
            Function<TEntity, TKey> dictionaryKeyGenerator,
            Function<TEntity, StringBuilder> insertCommandGenerator,
            Function<TEntity, StringBuilder> updateCommandGenerator,
            Function<TEntity, StringBuilder> deleteCommandGenerator,
            String insertHeader) {
        this.DictionaryKeyGenerator = dictionaryKeyGenerator;
        this.InsertCommandGenerator = insertCommandGenerator;
        this.UpdateCommandGenerator = updateCommandGenerator;
        this.DeleteCommandGenerator = deleteCommandGenerator;
        this.InsertHeader = insertHeader;
    }

    public final Function<TEntity, TKey> DictionaryKeyGenerator;
    public final Function<TEntity, StringBuilder> InsertCommandGenerator;
    public final Function<TEntity, StringBuilder> UpdateCommandGenerator;
    public final Function<TEntity, StringBuilder> DeleteCommandGenerator;

    /** "insert into &lt;table&gt; (col1,...) values " — replaces the legacy StaticExtInsertColumnParsedDic. */
    protected final String InsertHeader;

    public String EntityOrTableName() {
        // C# returned typeof(TEntity).Name, which is unavailable under Java generic erasure. This base
        // version is never invoked in practice (the only concrete subclass, SummaryCache, overrides it).
        throw new UnsupportedOperationException(
                "typeof(TEntity).Name is unavailable under Java generic erasure; subclasses must override EntityOrTableName().");
    }

    protected final ConcurrentMap<TKey, TEntity> Cache = new ConcurrentHashMap<>();
    protected final ConcurrentMap<TKey, TEntity> InsertedItems = new ConcurrentHashMap<>();
    protected final ConcurrentMap<TKey, TEntity> UpdatedItems = new ConcurrentHashMap<>();
    protected final ConcurrentMap<TKey, TEntity> DeletedItems = new ConcurrentHashMap<>();

    public boolean Exists(TKey key) {
        return this.Cache.containsKey(key);
    }

    public CachedItem<TKey, TEntity> GetItemByKey(TKey key) {
        TEntity selectedItem = this.Cache.get(key);
        return selectedItem == null ? null : new CachedItem<TKey, TEntity>(key, selectedItem);
    }

    public List<TEntity> GetItems() {
        return new ArrayList<>(this.Cache.values());
    }

    public List<TEntity> GetInsertedItems() {
        return new ArrayList<>(this.InsertedItems.values());
    }

    public List<TEntity> GetUpdatedItems() {
        return new ArrayList<>(this.UpdatedItems.values());
    }

    public List<TEntity> GetDeletedItems() {
        return new ArrayList<>(this.DeletedItems.values());
    }

    public int NumberOfInsertedItems() {
        return this.InsertedItems.size();
    }

    public int NumberOfUpdatedItems() {
        return this.UpdatedItems.size();
    }

    public int NumberOfDeletedItems() {
        return this.DeletedItems.size();
    }

    /**
     * Seed the cache with an EXISTING (already-persisted) row — the PopulatePrevSummary entry. It goes into
     * the cache only, NOT InsertedItems, so a later merge marks it Updated (not Inserted).
     */
    public void PopulateExisting(TEntity existing) {
        synchronized (this.locker) {
            TKey key = this.DictionaryKeyGenerator.apply(existing);
            if (this.Cache.putIfAbsent(key, existing) != null)
                throw new RuntimeException("Could not add existing item to cache while populating, probably duplicate item.");
        }
    }

    public void Delete(TKey key) {
        synchronized (this.locker) {
            TEntity toBeDeleted = this.Cache.get(key);
            if (toBeDeleted == null) throw new RuntimeException("Item does not exist in cache.");
            if (this.DeletedItems.containsKey(key) == false) {
                if (this.DeletedItems.putIfAbsent(key, toBeDeleted) != null)
                    throw new RuntimeException("Could not add item to deleted items.");
            }
            this.Cache.remove(this.DictionaryKeyGenerator.apply(toBeDeleted));
        }
    }

    public void UpdateThroughCache(TKey key, Consumer<TEntity> entityUpdater) {
        synchronized (this.locker) {
            if (this.DeletedItems.containsKey(key) == true)
                throw new RuntimeException("Item to be updated already exists in deleted items which is not allowed");
            TEntity entityToBeUpdated = this.Cache.get(key);
            if (entityToBeUpdated == null) throw new RuntimeException("Item does not exist in cache.");
            entityUpdater.accept(entityToBeUpdated);
            if (this.InsertedItems.containsKey(key) == false && this.UpdatedItems.containsKey(key) == false) {
                if (this.UpdatedItems.putIfAbsent(key, entityToBeUpdated) != null)
                    throw new RuntimeException("Could not add item to updated items, probably duplicate item.");
            }
        }
    }

    public CachedItem<TKey, TEntity> Insert(TEntity newItem, Predicate<TEntity> pkValidationMethod) {
        synchronized (this.locker) {
            TKey key = this.DictionaryKeyGenerator.apply(newItem);
            return InsertWithKey(newItem, key, pkValidationMethod);
        }
    }

    public CachedItem<TKey, TEntity> InsertWithKey(TEntity newItem, TKey key, Predicate<TEntity> pkValidationMethod) {
        synchronized (this.locker) {
            if (pkValidationMethod.test(newItem) == false)
                throw new RuntimeException("Primary key is either missing or invalid.");
            if (this.UpdatedItems.containsKey(key) == true)
                throw new RuntimeException("Item to be inserted already exists in updated items which is not allowed.");
            if (this.DeletedItems.containsKey(key) == true)
                throw new RuntimeException("Item to be inserted already exists in deleted items which is not allowed.");
            if (this.Cache.putIfAbsent(key, newItem) != null)
                throw new RuntimeException("Could not add item to cache, probably duplicate item.");
            if (this.InsertedItems.putIfAbsent(key, newItem) != null)
                throw new RuntimeException("Could not add item to inserted items, probably duplicate item.");
            return new CachedItem<TKey, TEntity>(key, newItem);
        }
    }

    // ---- write engine: thin executor over the single shared connection ----

    public void WriteAllChanges(ISqlExecutor executor) {
        WriteAllChanges(executor, BatchSqlWriter.DefaultSegmentSize);
    }

    public void WriteAllChanges(ISqlExecutor executor, int segmentSize) {
        synchronized (this.locker) {
            WriteDeletes(executor, segmentSize);
            WriteInserts(executor, segmentSize);
            WriteUpdates(executor, segmentSize);
        }
    }

    public int WriteInserts(ISqlExecutor executor) {
        return WriteInserts(executor, BatchSqlWriter.DefaultSegmentSize);
    }

    public int WriteInserts(ISqlExecutor executor, int segmentSize) {
        synchronized (this.locker) {
            if (this.InsertedItems.isEmpty() == true) return 0;
            if (this.InsertCommandGenerator == null) throw new RuntimeException("InsertCommandGenerator is not set and null.");
            List<StringBuilder> values = this.GetInsertedItems().stream()
                    .map(c -> this.InsertCommandGenerator.apply(c)).collect(Collectors.toList());
            int affected = BatchSqlWriter.WriteInsertsInSegments(executor, this.InsertHeader, values, segmentSize);
            this.InsertedItems.clear();
            return affected;
        }
    }

    public void WriteUpdates(ISqlExecutor executor) {
        WriteUpdates(executor, BatchSqlWriter.DefaultSegmentSize);
    }

    public void WriteUpdates(ISqlExecutor executor, int segmentSize) {
        synchronized (this.locker) {
            if (this.UpdatedItems.isEmpty() == true) return;
            if (this.UpdateCommandGenerator == null) throw new RuntimeException("UpdateCommandGenerator is null/not defined.");
            List<String> statements = this.GetUpdatedItems().stream()
                    .map(c -> this.UpdateCommandGenerator.apply(c).toString()).collect(Collectors.toList());
            BatchSqlWriter.WriteStatementsInSegments(executor, statements, segmentSize);
            this.UpdatedItems.clear();
        }
    }

    public void WriteDeletes(ISqlExecutor executor) {
        WriteDeletes(executor, BatchSqlWriter.DefaultSegmentSize);
    }

    public void WriteDeletes(ISqlExecutor executor, int segmentSize) {
        synchronized (this.locker) {
            if (this.DeletedItems.isEmpty() == true) return;
            if (this.DeleteCommandGenerator == null) throw new RuntimeException("DeleteCommandGenerator is null/not defined.");
            List<String> statements = this.GetDeletedItems().stream()
                    .map(c -> this.DeleteCommandGenerator.apply(c).toString()).collect(Collectors.toList());
            BatchSqlWriter.WriteStatementsInSegments(executor, statements, segmentSize);
            this.DeletedItems.clear();
        }
    }
}
